linq4j/src/main/java/org/apache/calcite/linq4j/MergeUnionEnumerator.java

```diff
@@ -0,0 +1,206 @@
+ * limitations under the License.
+ */
+package org.apache.calcite.linq4j;
+
+import org.apache.calcite.linq4j.function.EqualityComparer;
+import org.apache.calcite.linq4j.function.Function1;
+
+import org.checkerframework.checker.initialization.qual.UnknownInitialization;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.checkerframework.checker.nullness.qual.RequiresNonNull;
+
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Set;
+
+/**
+ * Performs a union (or union all) of all its inputs (which must be already sorted),
+ * respecting the order.
+ * @param <TSource> record type
+ * @param <TKey> sort key
+ */
+final class MergeUnionEnumerator<TSource, TKey> implements Enumerator<TSource> {
+  private final Enumerator<TSource>[] inputs;
+  private final TSource[] currentInputsValues;
+  private final boolean[] inputsFinished;
+  private final Function1<TSource, TKey> sortKeySelector;
+  private final Comparator<TKey> sortComparator;
+  private TSource currentValue;
+  private int activeInputs;
+
+  // Set to control duplicates, only used if "all" is false
+  private final @Nullable Set<EnumerableDefaults.Wrapped<TSource>> processed;
+  private final @Nullable Function1<TSource, EnumerableDefaults.Wrapped<TSource>> wrapper;
+  private @Nullable TKey currentKeyInProcessedSet;
+
+  private static final Object NOT_INIT = new Object();
+
+  MergeUnionEnumerator(
+      List<Enumerable<TSource>> sources,
+      Function1<TSource, TKey> sortKeySelector,
+      Comparator<TKey> sortComparator,
+      boolean all,
+      EqualityComparer<TSource> equalityComparer) {
+    this.sortKeySelector = sortKeySelector;
+    this.sortComparator = sortComparator;
+
+    if (all) {
+      this.processed = null;
+      this.wrapper = null;
+    } else {
+      this.processed = new HashSet<>();
+      this.wrapper = EnumerableDefaults.wrapperFor(equalityComparer);
+    }
+
+    final int size = sources.size();
+    //noinspection unchecked
+    this.inputs = new Enumerator[size];
+    int i = 0;
+    for (Enumerable<TSource> source : sources) {
+      this.inputs[i++] = source.enumerator();
+    }
+
+    //noinspection unchecked
+    this.currentInputsValues = (TSource[]) new Object[size];
+    this.activeInputs = this.currentInputsValues.length;
+    this.inputsFinished = new boolean[size];
+    //noinspection unchecked
+    this.currentValue = (TSource) NOT_INIT;
+
+    initEnumerators();
+  }
+
+  @RequiresNonNull("inputs")
+  @SuppressWarnings("method.invocation.invalid")
+  private void initEnumerators(@UnknownInitialization MergeUnionEnumerator<TSource, TKey> this) {
+    for (int i = 0; i < inputs.length; i++) {
+      moveEnumerator(i);
+    }
+  }
+
+  private void moveEnumerator(int i) {
+    final Enumerator<TSource> enumerator = inputs[i];
+    if (!enumerator.moveNext()) {
+      activeInputs--;
+      inputsFinished[i] = true;
```

Comment:
A slight note: it might be a good time to perform `currentInputsValues[i] = null` to release memory.

Reply:
Good point. I'll apply it.

Reply:
Doing so raises a problem in CheckerFramework:
```
C:\work\calcite_gradle\calcite\linq4j\src\main\java\org\apache\calcite\linq4j\MergeUnionEnumerator.java:102: error: [assignment.type.incompatible] incompatible types in assignment.
      currentInputsValues[i] = null;
                               ^
  found   : @Initialized @Nullable null
  required: TSource[ extends @Initialized @Nullable Object super @Initialized @NonNull Void]
```
@vlsi is there a way to specify that the elements inside `TSource[] currentInputsValues` can be `null`? Or should we suppress warning in this case?

Reply:
That depends.
a) The declaration can to be `@Nullable TSource[] currentInputsValues`. That would need to handle `null` in all use cases
b) `null` store could be suppressed

Reply:
I'd lead towards `b` in this particular case

Reply:
However, I am not being able to suppress the warning just in that specific line (the following suppression fails):
```
@SuppressWarnings("assignment.type.incompatible")
currentInputsValues[i] = null;
```
But it works fine when I add the suppression to the whole method (which I'd like to avoid in order to keep the suppression on the specific context where it must be applied). Any ideas @vlsi ?

Reply:
I'm afraid statement-level suppression works for `variable declaration statements` only :-/
So the WA could be to declare a variable that overrides the type for the cleanup purposes.

```java
@SuppressWarnings("...")
@Nullable TSouce inputsValues = currentInputsValues;
inputsValues[i] = null;
```

Or it could be method-level declaration (which I do not like either)
Or it could be "extract the clanup to a separate method with suppression".

Reply:
Thanks for the feedback, I applied your first WA.